inference-engine/src/gna_plugin/transformations/decompose_2d_conv.cpp

```diff
@@ -0,0 +1,582 @@
+
+struct GraphData {
+    std::shared_ptr<ngraph::opset7::Transpose>leading_transpose;
+    std::shared_ptr<ngraph::opset7::Convolution>conv;
+    std::shared_ptr<ngraph::opset7::Transpose>trailing_transpose;
+    std::shared_ptr<ngraph::op::util::UnaryElementwiseArithmetic>af;
+    std::shared_ptr<ngraph::opset7::MaxPool>max_pool;
+    std::shared_ptr<ngraph::Node>last_op_in_sequence_for_replacement;
+    std::shared_ptr<ngraph::Node>bias_const;
+    size_t conv_count;
+    size_t pool_size_width;
+    size_t pool_stride_width;
+    // TODO: currently 2D max pool is not supported
+    //size_t pool_size_height;
+    //size_t pool_stride_height;
+};
+
+static bool VerifyAndGetConvData(std::shared_ptr<ngraph::opset7::Convolution> conv, ConvData& conv_data) {
+    const auto& input = conv->input_value(0);
+    const auto& filters = conv->input_value(1);
+
+    // We support only batch == 1
+    if (input.get_shape()[0] != 1) {
+        return false;
+    }
+
+    size_t filter_height = filters.get_shape()[2];
+    size_t filter_width = filters.get_shape()[3];
+
+    if (filter_width > GNALimitations::copyMaxGrouping || filter_height > GNALimitations::copyMaxGrouping) {
+        return false;
+    }
+
+    GetConvData(conv, conv_data);
+
+    IE_ASSERT(conv_data.output_channel_count == conv->get_output_shape(0)[1]);
+
+    return true;
+}
+
+static std::shared_ptr<ngraph::Node> VerifyBiasAndReshapeConst(std::shared_ptr<ngraph::opset7::Add> conv_bias, const ConvData& conv_data) {
+    auto add_const = std::dynamic_pointer_cast<ngraph::op::Constant>(conv_bias->input_value(1).get_node_shared_ptr());
+
+    if (add_const) {
+        auto bias_size = shape_size(add_const->get_shape());
+
+        // The add may be a normal add not conv bias, then we just go further
+        if (bias_size == conv_data.filter_count) {
+            return ngraph::op::util::make_try_fold<ngraph::opset7::Reshape>(add_const,
+                ngraph::opset7::Constant::create(ngraph::element::i64, ngraph::Shape{4}, ngraph::Shape{1, bias_size, 1, 1}), false);
+        }
+    }
+    // Bias size does not match (or dynamic bias), can't decompose such convolution
+    return nullptr;
+}
+
+static bool VerifyMaxPool(GraphData& graph_data, std::shared_ptr<ngraph::opset7::MaxPool> max_pool) {
+    auto pool_filter = max_pool->get_kernel();
+    auto pool_strides = max_pool->get_strides();
+
+    // Check Max Pool padding and limitations
+    if (max_pool->get_auto_pad() != ngraph::op::PadType::VALID ||
+        pool_filter.size() != 2 || pool_strides.size() != 2 ||
+        pool_filter[0] > GNALimitations::maxPoolMaxWindowSize)
+        return false;
+
+    graph_data.pool_size_width = pool_filter[1];
+    graph_data.pool_stride_width = pool_strides[1];
+    return true;
+}
+
+static size_t CalculateConvCount(const ConvData& conv_data) {
+    // Check if split of plane due to GNA HW limitations of 768 filter elements is possible
+    size_t conv_count = 1;
```

Comment:
nit: may be it's better to set conv_count to 2 from the beginning instead of increasing it in the end?